### PR TITLE
Working on component-wise transformations that mimic `torch.cat` and `torch.stack`

### PR DESCRIPTION
**Pull request by ahmadsalim**
As discussed in #11755 .

### CODE REVIEW

**Review by vishwakftw — COMMENTED** [review]
Some minor comments. Pinging @alicanb and @fritzo for a review.

File: torch/distributions/transforms.py [hunk 1]
```diff
@@ -539,3 +541,83 @@ def _call(self, x):
...
+
+class CatTransform(Transform):
+    def __init__(self, tseq, dim=0):
+        assert all(isinstance(t, Transform) for t in tseq)
+        super(CatTransform, self).__init__()
+        self.transforms = list(tseq)
+        self.dim = dim
+    
+    def _call(self, x):
+        assert -len(x.shape) <= self.dim < len(x.shape)
```
[the remark this answers is not in the document]
Comment (ahmadsalim) [reply torch/distributions/transforms.py]: Thanks for the comments!

[the remark this answers is not in the document]
Comment (ahmadsalim) [reply torch/distributions/transforms.py]: I will take your improvements into account.

[the remark this answers is not in the document]
Comment (fritzo) [reply torch/distributions/transforms.py]: `-x.dim() <= self.dim < x.dim()` I believe

File: torch/distributions/constraints.py [hunk 2]
```diff
@@ -297,6 +301,41 @@ def check(self, value):
         return (value == value).all()  # False for NANs.
 
 
+class _Cat(Constraint):
+    """
+    Constraint functor that applies a sequence of constraints
+    `cseq` at the submatrixes at dimension `dim`
+    in a way compatible with `torch.cat`.
+    """
+    def __init__(self, cseq, dim=0):
```
Comment (fritzo) [inline torch/distributions/constraints.py]: nit: consider adding a line
```diff
  def __init__(self, cseq, dim=0):
+     cseq = list(cseq)
      ...
```
in case a user passes in a generator or mutable list for `cseq`

File: torch/distributions/constraints.py [hunk 3]
```diff
@@ -297,6 +301,41 @@ def check(self, value):
...
+                   for v, constr in zip(vs, self.cseq))
+
+
+class _Stack(Constraint):
+    """
+    Constraint functor that applies a sequence of constraints
+    `cseq` at the submatrixes at dimension `dim`
+    in a way compatible with `torch.stack`.
+    """
+    def __init__(self, cseq, dim=0):
```
Comment (fritzo) [inline torch/distributions/constraints.py]: ditto: `cseq = list(cseq)`

File: torch/distributions/constraints.py [hunk 4]
```diff
@@ -297,6 +301,41 @@ def check(self, value):
...
+    in a way compatible with `torch.cat`.
+    """
+    def __init__(self, cseq, dim=0):
+        assert all(isinstance(c, Constraint) for c in cseq)
+        self.cseq = cseq
+        self.dim = dim
+
+    def check(self, value):
+        assert -len(value.shape) <= self.dim < len(value.shape)
+        vs = value.chunk(value.shape[self.dim], self.dim)
```
Comment (fritzo) [inline torch/distributions/constraints.py]: See comment on `lengths` arg.

File: torch/distributions/transforms.py [hunk 5]
```diff
@@ -539,3 +541,112 @@ def _call(self, x):
     def _inverse(self, y):
         flat_y = y.contiguous().view((-1,) + y.shape[-2:])
         return torch.stack([self._inverse_on_event(z) for z in flat_y]).view(y.shape)
+
+
+class CatTransform(Transform):
+    """
+    Transform functor that applies a sequence of transforms `tseq`
+    component-wise to each submatrix at `dim` in a way compatible
+    with `torch.cat`.
```
Comment (fritzo) [inline torch/distributions/transforms.py]: nit: I think you can get sphinx to link to the docs via
```py
'''
Transform functor ... compatible with :func:`torch.cat`.
'''
```

File: torch/distributions/transforms.py [hunk 6]
```diff
@@ -539,3 +541,112 @@ def _call(self, x):
...
+    """
+    def __init__(self, tseq, dim=0):
+        assert all(isinstance(t, Transform) for t in tseq)
+        super(CatTransform, self).__init__()
+        self.transforms = list(tseq)
+        self.dim = dim
+    
+    def _call(self, x):
+        assert -x.dim() <= self.dim < x.dim()
+        assert x.size(self.dim) == len(self.transforms)
```
Comment (fritzo) [inline torch/distributions/transforms.py]: This assumes the `.size(self.dim)` of each input is tensor is 1. I think to handle a more general case you could add a `lengths` arg, maybe like
```py
def __init__(self, tseq, dim=0, lengths=None):
    if sizes is None:
        lengths = [1] * len(t_seq)
    assert len(lengths) == len(tseq)
    self.lengths = lengths
    ...

@lazy_property
def length(self):
    return sum(self.lengths)

def _call(self, x):
    ...
    assert x.size(self.dim) == self.length
    yslices = []
    start = 0
    for trans, length in zip(self.transforms, self.lengths):
        xslice = x.narrow(self.dim, start, size)
        yslices.append(trans(xslice))
        start = start + length  # take care to avoid += for jit compatibility
    ...
```

File: torch/distributions/constraint_registry.py [hunk 7]
```diff
@@ -212,3 +212,19 @@ def _transform_to_simplex(constraint):
 @transform_to.register(constraints.lower_cholesky)
 def _transform_to_lower_cholesky(constraint):
     return transforms.LowerCholeskyTransform()
+
+
+@biject_to.register(constraints.cat)
+@transform_to.register(constraints.cat)
+def _transform_to_cat(constraint):
+    return transforms.CatTransform([transform_to(c)
```
Comment (fritzo) [inline torch/distributions/constraint_registry.py]: See comment about `lengths` arg.

**Review by fritzo — COMMENTED** [review]
This is really nice! I especially like the `constraints.cat` and `.stack` :smile:

**Comment by ahmadsalim** [comment]
Thanks for the comments! I will take a look at generalizing for `length` as soon as possible.

File: torch/distributions/constraints.py [hunk 8]
```diff
@@ -297,6 +301,41 @@ def check(self, value):
...
+    in a way compatible with `torch.stack`.
+    """
+    def __init__(self, cseq, dim=0):
+        assert all(isinstance(c, Constraint) for c in cseq)
+        self.cseq = cseq
+        self.dim = dim
+
+    def check(self, value):
+        assert -len(value.shape) <= self.dim < len(value.shape)
+        vs = [value.select(self.dim, i) for i in range(value.shape[self.dim])]
```
Comment (vishwakftw) [inline torch/distributions/constraints.py]: Same as above.

File: torch/distributions/transforms.py [hunk 9]
```diff
@@ -539,3 +541,112 @@ def _call(self, x):
     def _inverse(self, y):
         flat_y = y.contiguous().view((-1,) + y.shape[-2:])
```
Comment (vishwakftw) [inline torch/distributions/transforms.py]: Please use `y.reshape((-1,) + y.shape[-2:])`

File: torch/distributions/transforms.py [hunk 10]
```diff
@@ -539,3 +541,112 @@ def _call(self, x):
...
+    def bijective(self):
+        return all(t.bijective for t in self.transforms)
+
+    @constraints.dependent_property
+    def domain(self):
+        return constraints.stack([t.domain for t in self.transforms], self.dim)
+
+    @constraints.dependent_property
+    def codomain(self):
+        return constraints.stack([t.codomain for t in self.transforms], self.dim)
```
Comment (vishwakftw) [inline torch/distributions/transforms.py]: No newline is causing lint to fail.

File: torch/distributions/constraints.py [hunk 11]
```diff
@@ -317,3 +356,5 @@ def check(self, value):
 lower_triangular = _LowerTriangular()
 lower_cholesky = _LowerCholesky()
 positive_definite = _PositiveDefinite()
+cat = _Cat
+stack = _Stack
```
Comment (vishwakftw) [inline torch/distributions/constraints.py]: No newline is causing lint to fail.

**Review by vishwakftw — COMMENTED** [review]
Sorry about not being clear earlier. Most of the changes recommended are pedantic ones, and some to fix lint failures.

**Comment by ahmadsalim** [comment]
@vishwakftw Thanks! I will fix the linting issues and the shape calls.

**Comment by ahmadsalim** [comment]
@fritzo @vishwakftw I have now taken all comments into account.
Could you kindly check again if there are any other necessary changes?

Thanks! 😄

File: torch/distributions/constraints.py [hunk 12]
```diff
@@ -297,6 +301,59 @@ def check(self, value):
...
+        return torch.cat(checks, self.dim)
+
+
+class _Stack(Constraint):
+    """
+    Constraint functor that applies a sequence of constraints
+    `cseq` at the submatrices at dimension `dim`,
+    each of size `lengths[dim]` in a way compatible with :func:`torch.stack`.
+    """
+    def __init__(self, cseq, dim=0, lengths=None):
```
Comment (fritzo) [inline torch/distributions/constraints.py]: Hmm I think `lengths` makes sense for `Cat` but not for `Stack`, right? IIUC `torch.stack` creates a new dimension, so the size of each thing in the new dimension will be 1.

File: torch/distributions/transforms.py [hunk 13]
```diff
@@ -533,9 +535,159 @@ def _inverse_on_event(self, y):
...
+                               self.dim, self.lengths)
+
+
+class StackTransform(Transform):
+    """
+    Transform functor that applies a sequence of transforms `tseq`
+    component-wise to each submatrix at `dim`, of size `lengths[dim]`
+    in a way compatible with :func:`torch.stack`.
+    """
+    def __init__(self, tseq, dim=0, lengths=None):
```
Comment (fritzo) [inline torch/distributions/transforms.py]: ditto: I think `lengths` makes sense for `Cat` but not for `Stack`. Am I missing something?

File: test/test_distributions.py [hunk 14]
```diff
@@ -4025,6 +4025,76 @@ def f(x):
             self.assertEqual(f(x), traced_f(x))
 
 
+class TestFunctors(TestCase):
+    def test_cat_transform(self):
```
Comment (fritzo) [inline test/test_distributions.py]: It would be good to add one more `test_cat_transform_lengths` with an example with nonuniform lengths. It's nice to test both code paths: uniform lengths (with lengths not provided) and nonuniform lengths.

File: torch/distributions/constraints.py (hunk 12 above)
Comment (ahmadsalim) [reply torch/distributions/constraints.py]: I think I agree! I was just thinking about how weird it was when I was doing the slicing.

File: test/test_distributions.py (hunk 14 above)
Comment (ahmadsalim) [reply test/test_distributions.py]: Sure, will do!

**Comment by ahmadsalim** [comment]
OK, everything should hopefully be fixed now 😄 !

File: torch/distributions/constraint_registry.py [hunk 15]
```diff
@@ -212,3 +212,20 @@ def _transform_to_simplex(constraint):
 @transform_to.register(constraints.lower_cholesky)
 def _transform_to_lower_cholesky(constraint):
     return transforms.LowerCholeskyTransform()
+
+
+@biject_to.register(constraints.cat)
```
Comment (fritzo) [inline torch/distributions/constraint_registry.py]: Oh one more thing, you'll want to register bijections and transforms separately:
```py
@biject_to.register(constraints.cat)
def _biject_to_cat(constraint):
    return transforms.CatTransform([biject_to(c) for c in constraint.cseq],
                                   constraint.dim,
                                   constraint.lengths)

@transform_to.register(constraints.cat)
def _transform_to_cat(constraint):
    return transforms.CatTransform([transform_to(c) for c in constraint.cseq],
                                   constraint.dim,
                                   constraint.lengths)
```
and similarly for `StackTransform`

File: torch/distributions/constraints.py [hunk 16]
```diff
@@ -297,6 +301,49 @@ def check(self, value):
         return (value == value).all()  # False for NANs.
 
 
+class _Cat(Constraint):
+    """
+    Constraint functor that applies a sequence of constraints
+    `cseq` at the submatrices at dimension `dim`,
+    each of size `lengths[dim]`, in a way compatible with :func:`torch.cat`.
+    """
```
Comment (fritzo) [inline torch/distributions/constraints.py]: nit: it might be nice to show code examples of how to use these cool new transforms.

Comment (fritzo) [reply torch/distributions/constraints.py]: I'd also recommend `make -C docs html` and checking that the docs look ok

**Comment by ahmadsalim** [comment]
@fritzo I added a few examples (which I also tested manually), and separated `biject_to` from `transform_to` as requested 😄 .

I unfortunately had issues with running `make -C docs html` with sphinx complaining about it couldn't find `pytorch`. Any idea why this happens?

**Review by fritzo — APPROVED** [review]
Looks great! I guess docs are broken at the moment 😕

**Comment by fritzo** [comment]
LGTM. @apaszke do you have any additional comments?

**Comment by ahmadsalim** [comment]
Any updates on the status of the PR? Thanks!

**Comment by ahmadsalim** [comment]
I am closing the PR since there doesn't seem to be any updates/responses on its status. I can reopen it if people are interested.

**Comment by ezyang** [comment]
We should merge this, sorry it has fallen on the wayside.

**Comment by ahmadsalim** [comment]
@ezyang Great, thanks for the update! 😄 .

**Comment by fritzo** [comment]
Hi @ahmadsalim sorry this languished. Could you merge with master and we can try to merge?

**Comment by ahmadsalim** [comment]
@fritzo Thanks for getting back to me on this PR!
I was trying to rebase it manually now, but luckily @SsnL used the PyTorch bot instead 😄

**Comment by ahmadsalim** [comment]
@fritzo It seems that the PR was rebased succesfully by the Pytorch bot 😄. Could you kindly take a look again when you have time? Thanks!

**Comment by fritzo** [comment]
Thanks for updating, LGTM.
